internal/controller/timedcommitstatus_controller.go

```diff
@@ -0,0 +1,434 @@
+		}
+		tcs.Status.Environments = append(tcs.Status.Environments, envTimedStatus)
+
+		// Create or update the CommitStatus for the current environment's active SHA
+		// This acts as an active commit status that gates promotions from this environment
+		err := r.upsertCommitStatus(ctx, tcs, ps, envConfig.Branch, currentActiveSha, phase, message, envConfig.Branch)
+		if err != nil {
+			return false, fmt.Errorf("failed to upsert CommitStatus for environment %q: %w", envConfig.Branch, err)
+		}
+
+		logger.Info("Processed environment time gate",
+			"branch", envConfig.Branch,
+			"activeSha", currentActiveSha,
+			"phase", phase,
+			"elapsed", elapsed.Round(time.Second),
+			"required", envConfig.Duration.Duration)
+	}
+
+	return timeGateTransitioned, nil
+}
+
+// calculateCommitStatusPhase determines the commit status phase based on time elapsed since deployment
+func (r *TimedCommitStatusReconciler) calculateCommitStatusPhase(commitTime time.Time, requiredDuration time.Duration, elapsed time.Duration, envBranch string) (promoterv1alpha1.CommitStatusPhase, string) {
+	if commitTime.IsZero() {
+		// If we don't have a commit time, we can't make a decision
+		return promoterv1alpha1.CommitPhasePending, fmt.Sprintf("Waiting for commit time to be available in %s environment", envBranch)
+	}
+
+	if elapsed >= requiredDuration {
+		// Sufficient time has passed
+		return promoterv1alpha1.CommitPhaseSuccess, fmt.Sprintf("Time-based gate requirement met for %s environment", envBranch)
+	}
+
+	// Not enough time has passed yet
+	return promoterv1alpha1.CommitPhasePending, fmt.Sprintf("Waiting for time-based gate on %s environment", envBranch)
+}
+
+func (r *TimedCommitStatusReconciler) upsertCommitStatus(ctx context.Context, tcs *promoterv1alpha1.TimedCommitStatus, ps *promoterv1alpha1.PromotionStrategy, branch, sha string, phase promoterv1alpha1.CommitStatusPhase, message string, envBranch string) error {
+	// Generate a consistent name for the CommitStatus
+	commitStatusName := utils.KubeSafeUniqueName(ctx, fmt.Sprintf("%s-%s-timed", tcs.Name, branch))
+
+	commitStatus := promoterv1alpha1.CommitStatus{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      commitStatusName,
+			Namespace: tcs.Namespace,
+		},
+	}
+
+	// Create or update the CommitStatus
+	_, err := ctrl.CreateOrUpdate(ctx, r.Client, &commitStatus, func() error {
+		// Set owner reference to the TimedCommitStatus
+		if err := ctrl.SetControllerReference(tcs, &commitStatus, r.Scheme); err != nil {
+			return fmt.Errorf("failed to set controller reference: %w", err)
+		}
+
+		// Set labels for easy identification
+		if commitStatus.Labels == nil {
+			commitStatus.Labels = make(map[string]string)
+		}
+		commitStatus.Labels["promoter.argoproj.io/timed-commit-status"] = utils.KubeSafeLabel(tcs.Name)
+		commitStatus.Labels["promoter.argoproj.io/environment"] = utils.KubeSafeLabel(branch)
+		commitStatus.Labels["promoter.argoproj.io/commit-status"] = "timed"
```

Comment:
We should probably document these labels as best practices for commit status controllers.